compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposePanel.desktop.kt

```diff
@@ -15,255 +15,29 @@
  */
 package androidx.compose.ui.awt
 
-import androidx.compose.runtime.Composable
-import androidx.compose.runtime.CompositionLocalProvider
+import androidx.compose.runtime.InternalComposeApi
 import androidx.compose.ui.ExperimentalComposeUiApi
-import androidx.compose.ui.focus.FocusDirection
-import androidx.compose.ui.window.WindowExceptionHandler
-import java.awt.*
-import java.awt.event.FocusEvent
-import java.awt.event.FocusListener
-import javax.swing.JLayeredPane
-import javax.swing.SwingUtilities.isEventDispatchThread
-import org.jetbrains.skiko.ClipComponent
-import org.jetbrains.skiko.GraphicsApi
 import org.jetbrains.skiko.SkiaLayerAnalytics
 
 /**
  * ComposePanel is a panel for building UI using Compose for Desktop.
  *
+ * It disposes its content on detach from Swing hierarchy.
+ * For manual dispose control see [DisposableComposePanel].
+ *
  * @param skiaLayerAnalytics Analytics that helps to know more about SkiaLayer behaviour.
  * SkiaLayer is underlying class used internally to draw Compose content.
  * Implementation usually uses third-party solution to send info to some centralized analytics gatherer.
  */
+@OptIn(InternalComposeApi::class)
 class ComposePanel @ExperimentalComposeUiApi constructor(
-    private val skiaLayerAnalytics: SkiaLayerAnalytics,
-) : JLayeredPane() {
+    skiaLayerAnalytics: SkiaLayerAnalytics,
+) : ComposePanelBase(skiaLayerAnalytics) {
```

Comment:
Moved conversation from https://github.com/JetBrains/compose-multiplatform-core/pull/620#issuecomment-1620309379

> I'm not a big believer in "experts" and their generic advice.

I am too :). But this is a valid advice in general from my experience, and it is not about all booleans. Different `boolean`'s can mean different things. If it affects behavior that can be separated, then it most probably should be just separated.

Examples of these booleans - `Text(editable)`, `ComposeWindow(isDialog)`, `LaunchedEffect(restartOnRecomposition = true)`, `Button(isOutlined)`, `clickable(forMouse)`, `TextField(openLinksOnClick)`, `JFrame.disposeOnClose` (existing one). Some of them are justified, some are not - it is often subjective.

`enabled` doesn't add completely different behavior, it just adds "perpendicular" dimension to the existing behavior.

Regarding `autoDisposable` - I am not completely against it, but it alters (doesn't extends) the current API/behavior of `ComposePanel` when we pass `autoDisposable  = false`:
- we need a new `dispose` method, that only makes sense if `autoDisposable = false`.
- clients of `ComposePanel` needs to know if the panel auto disposable (to make decision to call `dispose`). Imagine a framework that takes ComposePanel:
```
fun main() {
   val composePanel = ...
   initFramework(composePanel )
}
```
framework doesn't know that it needs to call `dispose`

Reply:
Exactly, because this autoDisposable is not exposed, its internal state (I think this constructor should be even private). Clients provide panelScope or it will be disposed automatically

Reply:
> `enabled` doesn't add completely different behavior, it just adds "perpendicular" dimension to the existing behavior.
I really don't see anything wrong with `autoDispose` here. I'm not familiar with Robert C. Martin's tip 12, but maybe he meant something else.

> Imagine a framework that takes ComposePanel:
There are several cases:
1. The framework decides whether the panel needs to be disposed. Then it either sets the `autoDispose` flag itself, or calls `dispose` itself, whatever is more convenient (calling dispose a second time should be ignored).
2. The framework doesn't decide whether the panel needs to be disposed. Then it doesn't dispose and the decision is left with the creator of the panel.

Also, how would that framework look with the subclass implementation? Would it check the type of the panel with `is DisposableComposePanel` to know whether to dispose?

Reply:
Sorry, I reread the snippet, and realised that there is no public `dispose` and `autoDisposable: Boolean`.

Still, introducing `panelScope: CoroutineScope` means the same thing until we can implement the old behavior via `panelScope: CoroutineScope`.

I can be subjective here, so would accept any well thought decision.

Reply:
> Exactly, because of this autoDisposable is not exposed, its internal state (I think this constructor should be even private). Clients provide panelScope or it will be disposed automatically

Maybe I'm not understanding the `panelScope` thing, but it seems a very roundabout and strange way to achieve the same thing. Instead of just saying `autoDispose=true` I need to provide a coroutine scope that when it's cancelled the panel is disposed?

Reply:
> The framework decides whether the panel needs to be disposed. Then it either sets the autoDispose flag itself

Making it setter instead of constructor adds another disadvantages.

Reply:
> The framework doesn't decide whether the panel needs to be disposed

Let's suppose that the framework can create dialogs with this panel. So, the client can't decide when to dispose, only framework knows when, but it doesn't know should it.

> how would that framework look with the subclass implementation

Good question. It shouldn't check `is DisposableComposePanel` (it adds coupling) or `composePanel.isAutoDisposable` (it adds one more burden).

I think, the framework needs only one type of ComposePanel. And if it decides that it needs preserving state in the detached state, it should deprecate `createPanel: ComposePanel`, and request `createDisposablePanel: DisposableComposePanel`, and only then be responsible of calling of `dispose`.

Reply:
I've introduced new solution, please check